### ISSUE TITLE
Hyperflux Support

### ISSUE THREAD

**Issue by Faeranne** [openrazer/openrazer#543]
The Razer Mamba Hyperflux and Firefly Hyperflux seem to be considered a single device.

Mamba:
Product ID: RZ83-02480100
Model Number: RZ01-0248

Firefly:
Product ID: RZ83-02480100
Model Number: RZ02-0248

Firefly plugged in:
[lsusb -d 1532:0068 -v](https://pastebin.com/iBr7UF4J)
[ls -lah /dev/input/by-id/](https://pastebin.com/E3xP3PCY)

Mamba plugged in:
[lsusb -d 1532:0069 -v](https://pastebin.com/25qG6nb8)
[ls -lah /dev/input/by-id/](https://pastebin.com/TF7Zj0C5)

I have to repair my Windows VM, then I'll get pcap's and screenshots.

**Comment by pashtet04** [comment]
Could you add support of this device please?

**Comment by redaxmedia** [comment]
### Mamba

![mouse-00](https://user-images.githubusercontent.com/1835397/52511113-f4303b80-2bfe-11e9-887a-eadfb7a333ef.PNG)
![mouse-01](https://user-images.githubusercontent.com/1835397/52511116-f72b2c00-2bfe-11e9-917f-8d1f9ad94ce4.PNG)
![mouse-02](https://user-images.githubusercontent.com/1835397/52511122-fa261c80-2bfe-11e9-8ba2-c663949880f2.PNG)
![mouse-03](https://user-images.githubusercontent.com/1835397/52511127-001bfd80-2bff-11e9-88e0-bd5fcf959768.PNG)
![mouse-04](https://user-images.githubusercontent.com/1835397/52511129-01e5c100-2bff-11e9-88c4-c59979a7f783.PNG)
![mouse-05](https://user-images.githubusercontent.com/1835397/52511130-03af8480-2bff-11e9-9884-bc0ad5863434.PNG)

### Firefly

(most of the settings are disabled)

![mouse-pad-00](https://user-images.githubusercontent.com/1835397/52511177-3194c900-2bff-11e9-8dcf-7305d3738c04.PNG)
![mouse-pad-01](https://user-images.githubusercontent.com/1835397/52511182-335e8c80-2bff-11e9-9fef-bba205775faf.PNG)

**Comment by henryruhs** [comment]
# Mamba Hyperflux
Only mouse plugged in!
Download pcapng: [mouse.zip](https://github.com/openrazer/openrazer/files/2847832/mouse.zip)
## Settings
![mouse-key-settings-01](https://user-images.githubusercontent.com/1835397/52521439-10969d00-2c2c-11e9-9410-4c46276f69c5.PNG)
![mouse-key-settings-03](https://user-images.githubusercontent.com/1835397/52521442-155b5100-2c2c-11e9-93b6-955a3f6c70de.PNG)
![mouse-key-settings-02](https://user-images.githubusercontent.com/1835397/52521441-13918d80-2c2c-11e9-8c21-f8189b1f8894.PNG)
## Performance
![mouse-key-performance-01](https://user-images.githubusercontent.com/1835397/52521444-18564180-2c2c-11e9-94a5-b2f6b39192b2.PNG)
## Lightning
![mouse-key-lightning-01](https://user-images.githubusercontent.com/1835397/52521446-1ab89b80-2c2c-11e9-8b76-f82d6ef9df5d.PNG)
## Calibration
![mouse-key-calibratopm-01](https://user-images.githubusercontent.com/1835397/52521448-22784000-2c2c-11e9-887c-c09bc0e09ed0.PNG)
# Firefly Hyperflux
Only mouse-mat plugged in!
Download pcapng: [mouse-mat.zip](https://github.com/openrazer/openrazer/files/2847831/mouse-mat.zip)
## Lightning
![mouse-mat-lightning-01](https://user-images.githubusercontent.com/1835397/52521450-25733080-2c2c-11e9-95a9-225ba60bb178.PNG)

**Comment by hedmo** [comment]
@z3ntu what does milestone mean? i am on the hyperflux road to.i am running polychromatic-0.4.0 and seeing an unrecognized:1532:0068 in the menu.

regards hedmo

**Comment by z3ntu** [comment]
@hedmo Ignore the milestone, I use it more or less as a label for me to know what generation of Razer devices this is from

**Comment by hedmo** [comment]
@z3ntu .thanks for the reply .but how about the hyperflux ? is it going to  be supported at any time.i dont understand if it is or going to be supported..

regards

**Comment by z3ntu** [comment]
I'm lately busy with other projects so I don't have much time for OpenRazer. This device should be compared to other devices easy to implement, you could take a look at recent merged pull requests for a reference on what has to be done for the device to be supported.

**Comment by hedmo** [comment]
@z3ntu okay thanks.i understand. i am going try to help an add this .as i understand this is like the firefly and just need to change the id and stuff to make it work.maybe the hero @nightsky30 can help me this time to?.but for now i am going to find a pull requests for an added device

regards hedmo

**Comment by nightsky30** [comment]
@hedmo Like you said, maybe try adapting something like the the Firefly for the Firefly Hyperflux.  Looks like the Mamba Hyperflux could be similar to the Lancehead as wired/wireless options but without the side LEDs, and a capacitor vs a battery. Also not sure if the PID changes when the Mamba is plugged in vs unplugged.  Looks like it shares the Firefly cable.  So the mouse may only need a single class/device type?  You could use #887 as an example to work from, and tweak/remove what isn't needed.

**Comment by hedmo** [comment]
@nightsky30 .at the moment i am trying to use the firefly setup by patching : 
mouse_mat.py
razermousemat_driver.h
and
99-razer.rules
with the 0068 id.the device is recognized but it is not changing the rgbs so i think there is more files i need to patched.i will check #887 and i am going to see if i can find the commit that adds the firefly and work from there.

rerards

**Comment by hedmo** [comment]
@nightsky30 i have manage to use the Goliathus setup to work with firefly_hyperflux. the firefly setup did not work and i dont know why.i have edit the files to get the new product working at https://github.com/hedmo/openrazer .the problem now is that Goliathus and firefly are configured different and i dont know how to add wave to firefly_hyperflux
![firefly_hyperflux](https://user-images.githubusercontent.com/1720603/70894635-af7aeb00-1fed-11ea-8312-993fdfa29021.png)

regards hedmo

**Comment by nightsky30** [comment]
A couple things.  Not sure if the _Goliathus Chroma_ or _Goliathus Chroma Extended_ support wave through software or hardware, but from a video on Razer's site they appear to support the wave effect _somehow_.

BUT, regardless wave was not add for those devices in openrazer.  The Goliathus devices use the extended matrix effects and ZERO_LED vs the regular firefly that is using the standard matrix effects and BACKLIGHT_LED.

You can try adding the following and see if the firefly hyperflux supports the wave effect through hardware.

1) Add 'set_wave_effect' in the list of methods for the firefly hyperflux class in your version of mouse_mat.py.

2) Then I would try modifying **razer_attr_write_mode_wave** in razermousemat_driver.c so that it includes a switch statement to account for the different device types with a case statement for the firefly hyperflux, and a default statement for the normal firefly.  The firefly hyperflux appears to be using extended matrix effects with ZERO_LED.
```
    ...
    struct razer_report report = {0};

    switch (usb_dev->descriptor.idProduct) {
    case USB_DEVICE_ID_RAZER_FIREFLY_HYPERFLUX:        
        report = razer_chroma_extended_matrix_effect_wave(VARSTORE, ZERO_LED, direction);
        break;

    default:
        report = razer_chroma_standard_matrix_effect_wave(VARSTORE, BACKLIGHT_LED, direction);
        break;
    }

    razer_send_payload(usb_dev, &report);

    return count;
    ...
```

**Comment by hedmo** [comment]
@nightsky30 thanks for the hint but no go but now i can turn the light on and of ... i have set MATRIX_DIMS = [1, 15] in mouse_mat.py  but only one (the first) is working and lights all.

edit:i need MATRIX_DIMS = [1, 15] in mouse_mat.py to get all lights working

**Comment by hedmo** [comment]
@nightsky30 @z3ntu  i am stuck. i dont know where to go from here with the wave function.i will try to give you as much info as i can :

what works = spectrum,breath,static and brightness
what does now work= wave and in RazerGenie the matrix function= set color per zone 1-15(only zone 1 works and lights all).
the  MATRIX_DIMS = [1, 15] is the light zones as i understand? and that one should be set to MATRIX_DIMS = [1, 1] as i would like to have it but if i put it like that,the lights get fuzzy and only zone 1 will light up fully.

this is as far i have manage to get.it would be nice fix the last lose end (wave) to get this pulled
for other users to control the hyperflux and for me to move to the hyperflux mouse and the nari ultimate.i understand if time is a problem.

regards hedmo

**Comment by z3ntu** [comment]
As for effects, Synapse only uses the 'static' effects in hardware it seems, so I can't help you with wave.

For matrix dimensions, Synapse only uses row 0 column 0 start (& column 1 end), so that would match the `MATRIX_DIMS = [1, 1]`

**Comment by hedmo** [comment]
@z3ntu. okey as i understand,this is as good as it gets but i will try MATRIX_DIMS = [1, 1] one more time .then i like to pull my https://github.com/hedmo/openrazer to yours but i dont know how?

regards

**Comment by nightsky30** [comment]
@hedmo It seems like you cloned openrazer and pushed rather than forked from the official openrazer.  So **I think** the upstream connection is missing.  You could try adding the official openrazer repo as an upstream to your local master.  Then fetch --all and merge on your local master with the remote upstream.  And I think finally push to your openrazer remote repo before creating a pull request.

Otherwise, fork the official openrazer, clone your fork to a separate folder, create a branch, checkout said branch, reapply your changes to your forked local branch, then push and track the branch to your remote fork, and create a pull request from there.

Forgive me if I'm not clear enough or I had the wrong impression ^^

**Edit**:  Also, it is advisable to use separate branches for each device/pull request.  So perhaps forking might be best.

**Edit 2**:  Ah, never mind, looks like you got it figured out! :+1:

**Comment by hedmo** [comment]
@nightsky30 

yes i had a clone. this is new to me. i had to clone mine,delete my repo,fork openrazer,readd my clone to the repo. my clone was older then openrazer so i had to manually update the razer orbweaver chroma update.hoping to get help with my pull request #981

thank you for all the help

regards hedmo

**Comment by z3ntu** [comment]
Fixed with #981

**Comment by Nama** [comment]
I can't set the DPI of the Mamba Hyperflux. Shouldn't it be added with this? Only Firefly is showing up (which is prolly expected), but there aren't any DPI or button settings.
